### ISSUE TITLE
Is MKL being used?

### ISSUE THREAD

**Issue by crocha700** [pyqg/pyqg#112]
[landscape](https://landscape.io/github/crocha700/pyqg/19/messages/smell) complains mkl is not used. 

Apparently anaconda already builds numpy with mkl. I get the following message whenever importing numpy

```{python}
Vendor:  Continuum Analytics, Inc.
Package: mkl
Message: trial mode expires in 21 days
Vendor:  Continuum Analytics, Inc.
Package: mkl
Message: trial mode expires in 21 days
Vendor:  Continuum Analytics, Inc.
Package: mkl
Message: trial mode expires in 21 days
```

**Comment by rabernat** [comment]
I have personally never paid much attention to the mkl stuff. Now that we have so much of the heavy lifting in cython / pyfftw, I am not sure that numpy mkl optimizations have much effect. I would be happy to remove this complication.

Have you done any benchmarking with / without mkl?

**Comment by crocha700** [comment]
I haven't done any benchmarking. I agree that there's likely little benefits. Currently, this only use would be in the methods that use ```np.linalg```...but those are only called once when calculating the inversion matrix, vertical modes, etc. 

The point is, if MKL is installed with conda, then it is likely that numpy is already built with MKL -- no need to set ```fast_numpy```.

**Comment by MFJansen** [comment]
What about if people don't have pyfftw (but do have MKL)? Then it would make a huge difference whether default of MKL ffts are used!
Also: with few cores, I think MKL actually may often be faster than fftw. So it may not be bad to have that as an option. Maybe that's no longer true once cython is used though?

**Comment by rabernat** [comment]
> Also: with few cores, I think MKL actually may often be faster than fftw.

You might be comparing apples and oranges. Do you mean the the numpy.fft linked with MKL?

The MKL api is actually identical to fftw. In theory, it should be possible to build pyfftw directly with MKL.

**Comment by MFJansen** [comment]
> You might be comparing apples and oranges. Do you mean the the numpy.fft linked with MKL?
Yes. (Except in Canopy I think it's only fastnumpy that uses the MKL libraries). Can you elaborate what you mean when you say I'm comparing apples and oranges?

**Comment by rabernat** [comment]
> Can you elaborate what you mean when you say I'm comparing apples and oranges?

You are using the "numpy interface" to call MKL, which has different (larger) overhead than the pyfftw interface. I have no doubt that numpy with MKL is faster than numpy without MKL. It would be much more interesting to know
- if numpy with MKL is actually faster than pyfftw for a given number of threads
- if pyfftw linked against MKL is faster than pyfftw linked against FFTW (currently not an option, but possible with work)

**Comment by MFJansen** [comment]
>  I have no doubt that numpy with MKL is faster than numpy without MKL.

Wouldn't this be enough reason to keep MKL support in the model, keeping in mind that not everybody will have pyfftw?

> It would be much more interesting to know if numpy with MKL is actually faster than pyfftw for a given number of threads

I vaguely seem to recall that I did some crude tests indicating that they are at least comparable. But this was with MKL on one computer and pyfftw on the other one. So not possible to make an exact comparison. In either case, I personally won't obsess about small speed differences. More important to me seems that the model remains reasonably fast on a range of machines (including ones that don't have fftw), which brings me back to the first point.

**Comment by crocha700** [comment]
I guess my point was not to completely exclude MKL. I was just asking whether it is really necessary to ```import mkl``` and set ```numpy.fast_numpy = True```.  At least on my computer with anaconda, ```numpy``` is already built with MKL, and ```numpy.fast_numpy``` does not even exist here. 

```{python}
In [5]: np.version.version
Out[5]: '1.10.1'
```

**Comment by crocha700** [comment]
But ```numpy.use_fastnumpy``` does exist in the ```numpy``` version that came with ``canopy```

```{python}
In [1]: import numpy
In [2]: numpy.version.version
Out[2]: '1.9.2'
In [3]: numpy.use_fastnumpy
Out[3]: False
```
